app/src/main/java/dev/jdtech/jellyfin/viewmodels/HomeViewModel.kt

```diff
@@ -1,119 +1,112 @@
     private val application: Application,
-    private val jellyfinRepository: JellyfinRepository
+    private val repository: JellyfinRepository
 ) : ViewModel() {
 
+    private val views = MutableLiveData<List<HomeItem>>()
+    private val state = MutableSharedFlow<State>(
+        replay = 0,
+        extraBufferCapacity = 1,
+        onBufferOverflow = BufferOverflow.DROP_OLDEST
+    )
+
+    init {
+        loadData(updateCapabilities = true)
+    }
+
     private val continueWatchingString = application.resources.getString(R.string.continue_watching)
     private val nextUpString = application.resources.getString(R.string.next_up)
 
-    private val _views = MutableLiveData<List<HomeItem>>()
-    val views: LiveData<List<HomeItem>> = _views
+    fun views(): LiveData<List<HomeItem>> = views
 
-    private val _items = MutableLiveData<List<BaseItemDto>>()
-    val items: LiveData<List<BaseItemDto>> = _items
-
-    private val _finishedLoading = MutableLiveData<Boolean>()
-    val finishedLoading: LiveData<Boolean> = _finishedLoading
+    fun onStateUpdate(
+        scope: LifecycleCoroutineScope,
+        collector: (State) -> Unit
+    ) {
+        scope.launch { state.collect { collector(it) } }
+    }
 
-    private val _error = MutableLiveData<String>()
-    val error: LiveData<String> = _error
+    fun refreshData() = loadData(updateCapabilities = false)
 
-    init {
-        loadData()
-    }
+    private fun loadData(updateCapabilities: Boolean) {
+        state.tryEmit(Loading(inProgress = true))
 
-    fun loadData() {
-        _error.value = null
-        _finishedLoading.value = false
         viewModelScope.launch {
             try {
+                if (updateCapabilities) repository.postCapabilities()
 
-
-                jellyfinRepository.postCapabilities()
-
-                val items = mutableListOf<HomeItem>()
+                val updated = loadDynamicItems() + loadViews()
+                views.postValue(updated)
 
                 withContext(Dispatchers.Default) {
-
-                    val resumeItems = jellyfinRepository.getResumeItems()
-                    val resumeSection =
-                        HomeSection(UUID.randomUUID(), continueWatchingString, resumeItems)
-
-                    if (!resumeItems.isNullOrEmpty()) {
-                        items.add(HomeItem.Section(resumeSection))
-                    }
-
-                    val nextUpItems = jellyfinRepository.getNextUp()
-                    val nextUpSection = HomeSection(UUID.randomUUID(), nextUpString, nextUpItems)
-
-                    if (!nextUpItems.isNullOrEmpty()) {
-                        items.add(HomeItem.Section(nextUpSection))
-                    }
+                    syncPlaybackProgress(repository, application)
                 }
+            } catch (e: Exception) {
+                Timber.e(e)
+                state.tryEmit(LoadingError(e.toString()))
+            }
+            state.tryEmit(Loading(inProgress = false))
```

Comment:
The loading error is never shown because we emit Loading right after

Reply:
Good catch, thanks.